### PR TITLE
Add si32 HAL

### PR DESCRIPTION
**Pull request by rettichschnidi**
This is an update of #29, because I can not change it. It contains the latest version of #29 (98b458d99b74083fce4c715fedcee45d388eb8b8), but got rebased on the most recent main and includes a fixed assertion and optionally pulls in the relevant files when DMA and/or SPI is enabled.

This code is used by https://github.com/zephyrproject-rtos/zephyr/pull/76677.

### CODE REVIEW

**Review by jerome-pouiller — APPROVED** [review]

**Comment by rettichschnidi** [comment]
@jhedberg Repo growth is 6.3 % (11.86 MiB -> 12.61 MiB).

Current main:

```console
git clone --branch=main --single-branch https://github.com/zephyrproject-rtos/hal_silabs.git hal_silabs-upstream
$ git -C hal_silabs-upstream/ count-objects --verbose --human-readable 
count: 0
size: 0 bytes
in-pack: 4845
packs: 1
size-pack: 11.86 MiB
prune-packable: 0
garbage: 0
size-garbage: 0 bytes
```

Including the Si32 HAL:

```console
$ git clone --branch=gardena/rs/upstream/sim3u --single-branch git@github.com:husqvarnagroup/zephyrproject-rtos-hal_silabs.git hal_silabs-sim3u
$ git -C hal_silabs-sim3u/ count-objects --verbose --human-readable count: 0
size: 0 bytes
in-pack: 5176
packs: 1
size-pack: 12.61 MiB
prune-packable: 0
garbage: 0
size-garbage: 0 bytes
```

On the file system, the increase is also quite low (~4.2%): 

```console
$ du -sh *
4.0K    CMakeLists.txt
186M    gecko
7.8M    si32
76K     zephyr
```

**Review by asmellby — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

File: si32/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/tmp
```
Comment (jhedberg) [inline si32/.gitignore:1]: @rettichschnidi curious - why is this needed? Is something in the si32 SDK creating it?

Comment (rettichschnidi) [reply si32/.gitignore:1]: This folder gets created by `si32/scripts/extract_hal`.

Comment (rettichschnidi) [reply si32/.gitignore:2]: Just added a comment to the file.
